### ISSUE TITLE
Smarty::setExtensions() killing registered filters

### ISSUE THREAD

**Issue by SlowFox71** [smarty-php/smarty#1006]
Instructions to add output filters are:

```
$smarty->registerFilter("output", "protect_email");
```

Instructions to add custom extensions (overriding the defaults) are:

```
$smarty->setExtensions([
    new Smarty\Extension\CoreExtension(),
    new MyCustomExtension(),
    new Smarty\Extension\DefaultExtension(),
]);
```

Unfortunately the latter kills $smarty->BCPluginsAdapter thus disabling any filters. I solved the problem by using:

```
$extensions = $smarty->getExtensions();
array_splice($extensions, 1, 0, [new MyCustomExtension]);
$this->setExtensions($extensions);
```

I don't like the style, but at least it is working. But something must be done here, either in the API, or at least the documentation should be updated.

**Comment by wisskid** [comment]
Yes, if you also want the BCPluginsAdapter, you could also do this:

```php
$smarty->setExtensions([
    new Smarty\Extension\CoreExtension(),
    new MyCustomExtension(),
    new Smarty\Extension\DefaultExtension(),
   new Smarty\Extension\BCPluginsAdapter($smarty),
]);
```

I didn't document that because I thought people who write custom Extensions and insert them before the default ones probably don't need the BCPluginsAdapter anymore.

**Comment by SlowFox71** [comment]
Ok.

But am I really supposed to create (yet another) extension, just to register a filter? This seems kind of overkill to me.

**Comment by wisskid** [comment]
No, you don't need to. But if you are creating a custom extension anyway, why not put everything you need in there?

**Comment by SlowFox71** [comment]
The custom extension is created and included in my framework.

One single call in the application using that framework needs a special output filter, and it took me quite a while to figure out why it is ignored.

What else besides the (deprecated?) `registerFilter()` could be used for the filter?

**Comment by wisskid** [comment]
Add 
```php
	public function getOutputFilters(): array {
		return [ ];
	}
```
and return an object that implements `\Smarty\Filter\FilterInterface`. For example:

```php
class MyReverseFilter implements \Smarty\Filter\FilterInterface {

	public function filter($code, \Smarty\Template $template) {
		return strrev($code);
	}

}
```

Btw: you are right about the documentation. It needs more love.

**Comment by SlowFox71** [comment]
From a logical point of view MyCustomExtension is the wrong place. Perhaps I'll create a second, controller-specific extension, I'll have to think about it.
